crates/sui-core/src/transaction_input_loader.rs

```diff
@@ -0,0 +1,350 @@
+        for objref in receiving_objects {
+            // Note: the digest is checked later in check_transaction_input
+            let (object_id, version, _) = objref;
+
+            if self
+                .store
+                .have_received_object_at_version(object_id, *version, epoch_id)?
+            {
+                receiving_results.push(ReceivingObjectReadResult::new(
+                    *objref,
+                    ReceivingObjectReadResultKind::PreviouslyReceivedObject,
+                ));
+                continue;
+            }
+
+            let Some(object) = self.store.get_object(object_id)? else {
+                return Err(UserInputError::ObjectNotFound {
+                    object_id: *object_id,
+                    version: Some(*version),
+                }
+                .into());
+            };
+
+            receiving_results.push(ReceivingObjectReadResult::new(*objref, object.into()));
+        }
+
+        Ok((
+            input_results
+                .into_iter()
+                .map(Option::unwrap)
+                .collect::<Vec<_>>()
+                .into(),
+            receiving_results.into(),
+        ))
+    }
+
+    /// Reads input objects assuming a synchronous context such as the end of epoch transaction.
+    /// By "synchronous" we mean that it is safe to read the latest version of all shared objects,
+    /// as opposed to relying on the shared input version assignment.
+    #[instrument(level = "trace", skip_all)]
+    pub async fn read_objects_for_synchronous_execution(
+        &self,
+        tx_digest: &TransactionDigest,
+        input_object_kinds: &[InputObjectKind],
+        protocol_config: &ProtocolConfig,
+    ) -> SuiResult<InputObjects> {
+        self.read_objects_for_synchronous_execution_impl(
+            Some(tx_digest),
+            input_object_kinds,
+            &[],
+            protocol_config,
+        )
+        .await
+        .map(|r| r.0)
+    }
+
+    /// Read input objects for a dry run execution.
+    #[instrument(level = "trace", skip_all)]
+    pub async fn read_objects_for_dry_run_exec(
+        &self,
+        tx_digest: &TransactionDigest,
+        input_object_kinds: &[InputObjectKind],
+        receiving_objects: &[ObjectRef],
+        protocol_config: &ProtocolConfig,
+    ) -> SuiResult<(InputObjects, ReceivingObjects)> {
+        self.read_objects_for_synchronous_execution_impl(
+            Some(tx_digest),
+            input_object_kinds,
+            receiving_objects,
+            protocol_config,
+        )
+        .await
+    }
+
+    /// Read input objects for dev inspect
+    #[instrument(level = "trace", skip_all)]
+    pub async fn read_objects_for_dev_inspect(
+        &self,
+        input_object_kinds: &[InputObjectKind],
+        receiving_objects: &[ObjectRef],
+        protocol_config: &ProtocolConfig,
+    ) -> SuiResult<(InputObjects, ReceivingObjects)> {
```

Comment:
This function should not return ReceivingObjects today because it's effectively always empty, until we actually implement that part. We should make it so and add a TODO

Reply:
I started trying to remove this, and found that for dry-run, all the other machinery is already in place, so instead I changed this to read the receiving objects and return them.